nrf52-hal-common/src/saadc.rs

```diff
@@ -0,0 +1,111 @@
+//use crate::gpio::{Floating, Input, Pin};
+use crate::gpio::{Floating, Input};
+use crate::target::{saadc, SAADC};
+use core::ops::Deref;
+use embedded_hal::adc::{Channel, OneShot};
+
+// Only 1 channel is allowed right now, a discussion needs to be had as to how
+// multiple channels should work (See SCAN_MODE)
+
+pub trait SaadcExt: Deref<Target = saadc::RegisterBlock> + Sized {
+    fn constrain(self) -> Saadc;
+}
+
+pub struct Saadc(SAADC);
+
+impl Saadc {
+    pub fn new(saadc: SAADC) -> Self {
+        saadc.enable.write(|w| w.enable().enabled());
+        saadc.resolution.write(|w| w.val()._12bit());
+        saadc.oversample.write(|w| w.oversample().bypass());
+        saadc.samplerate.write(|w| w.mode().task());
+
+        saadc.ch[0].config.write(|w| {
+            w.refsel()
+                .internal()
+                .gain()
+                .gain1_6()
+                .tacq()
+                ._20us()
+                .mode()
+                .se()
+                .resp()
+                .bypass()
+                .resn()
+                .bypass()
+                .burst()
+                .enabled()
+        });
+        saadc.ch[0].pseln.write(|w| w.pseln().nc());
+
+        // Calibrate
+        saadc.tasks_calibrateoffset.write(|w| unsafe { w.bits(1) });
+        while saadc.events_calibratedone.read().bits() == 0 {}
+
+        Saadc(saadc)
+    }
+}
+
+impl<PIN> OneShot<Saadc, u16, PIN> for Saadc
+where
+    PIN: Channel<Saadc, ID = u8>,
+{
+    type Error = ();
+    fn read(&mut self, pin: &mut PIN) -> nb::Result<u16, Self::Error> {
+        match PIN::channel() {
+            0 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input0()),
+            1 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input1()),
+            2 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input2()),
+            3 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input3()),
+            4 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input4()),
+            5 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input5()),
+            6 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input6()),
+            7 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input7()),
+            // This can never happen as if another pin was used, there would be a compile time error
+            _ => panic!("Invalid pin"),
+        }
+
+        let mut val = 0u16;
+        self.0
+            .result
+            .ptr
+            .write(|w| unsafe { w.ptr().bits(((&mut val) as *mut _) as u32) });
```

Comment:
In other places of this module` compiler_fence(SeqCst);`  - compiler fence have been used to prevent optimizations as the compiler was not too aware of EasyDMA. Not sure if this way of getting the pointer will prevent this error or not. Could be worth looking into.

Reply:
I've added a fence between setting the pointer and max count, and starting and sampling the ADC, to ensure that we never try to sample into an old pointer. Does that sound correct?